lib/libp2phttp.go

```diff
@@ -0,0 +1,156 @@
+package vole
+
+import (
+	"context"
+	"crypto/ecdsa"
+	"crypto/elliptic"
+	"crypto/rand"
+	"crypto/tls"
+	"crypto/x509"
+	"crypto/x509/pkix"
+	"fmt"
+	"math/big"
+	"net"
+	"net/http"
+	"net/http/httputil"
+	"time"
+
+	"github.com/libp2p/go-libp2p/core/host"
+	"github.com/libp2p/go-libp2p/core/peer"
+	libp2phttp "github.com/libp2p/go-libp2p/p2p/http"
+	"github.com/multiformats/go-multiaddr"
+)
+
+func Libp2pHTTPSocketProxy(ctx context.Context, p multiaddr.Multiaddr, unixSocketPath string) error {
+	h, err := libp2pHost()
+	if err != nil {
+		return err
+	}
+
+	httpHost := libp2phttp.Host{StreamHost: h}
+
+	ai := peer.AddrInfo{
+		Addrs: []multiaddr.Multiaddr{p},
+	}
+	idStr, err := p.ValueForProtocol(multiaddr.P_P2P)
+	if err == nil {
+		id, err := peer.Decode(idStr)
+		if err != nil {
+			return err
+		}
+		ai.ID = id
+	}
+
+	hasTLS := false
+	hasHTTP := false
```

Comment:
> Could this logic get into trouble if say /ws/tls is being used for a relay and combined with /http?

`/http` represents the [HTTP transport](https://github.com/libp2p/specs/blob/master/http/transport-component.md). It shouldn't be combined with `/ws`.

> Maybe not relevant though since I'm not sure if there are multiaddrs where /http is valid when combined with /p2p/.

Not yet, but these could be combined. I just need to finish peer id auth over http and we could have a valid multiaddr that looks like `/.../http/p2p/12Foo`